### PR TITLE
My Site Dashboard: handle local and remote cards

### PR DESCRIPTION
**Pull request by leandroalonso**
Part of #17873

**Important: handling errors, persistence, update, pull to refresh, etc are not part of this task.**

This PR handles:

* Local cards
* Remote cards
* Configuring them

I will probably expand on the architecture in a future post, but the idea is quite simple: cards are registered in `DashboardCard` and they can be: 1) remote or 2) local. Remote cards are powered by remote data, while local cards aren't.

### To test

Make sure to run your sandbox with the 1.1 API diff applied and to have the My Site Dashboard flag enabled:

1. Open the app
2. Tap "Dashboard"
3. Notice that the Quick Actions row appear
4. Posts card might appear if 1) you have draft posts or 2) you have scheduled posts

For (4) you can play by creating drafts/scheduled posts and closing and reopening the app (updates are not handled yet).

## Regression Notes
1. Potential unintended areas of impact
n/a

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Added unit tests for the parser.

3. What automated tests I added (or what prevented me from doing so)
Added unit tests for the parser.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by leandroalonso** [comment]
@momo-ozawa sorry about the size of the PR, I recommend going commit by commit because that should make the review easier for you.

**Comment by leandroalonso** [comment]
@pachlava another technical one, assigning you for visibility. Feel free to test it. :)

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift
Comment (momo-ozawa) [inline WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift:18]: What do you think of defining a static var for `remoteCases` and `localCases` for the `DashboardCard` enum?

```
// DashboardCard.swift

static var remoteCases: [DashboardCard] {
    return DashboardCard.allCases.filter { $0.isRemote }
}

static var localCases: [DashboardCard] {
    return DashboardCard.allCases.filter { !$0.isRemote }
}
```

That way, we can avoid a giant if block and do something like 

```
remoteService.fetch(cards: cardsToFetch, forBlogID: wpComID, success: { [weak self] cards in

    DashboardCard.remoteCases.forEach { card in ... }

    DashboardCard.localCases.forEach { card in ... }
    
}

```

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift
Comment (momo-ozawa) [inline WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift]: I found this statement a bit hard to read. What do you think of something like this?
```suggestion
        let draftsCount = (posts["draft"] as? Array<Any>)?.count ?? 0
        let scheduledCount = (posts["scheduled"] as? Array<Any>)?.count ?? 0

        let hasDrafts = draftsCount > 0
        let hasScheduled = scheduledCount > 0
```

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/ViewModel/BlogDashboardViewModel.swift
Comment (momo-ozawa) [inline WordPress/Classes/ViewRelated/Blog/Blog Dashboard/ViewModel/BlogDashboardViewModel.swift]: I think we can get rid of this

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/ViewModel/BlogDashboardViewModel.swift
Comment (momo-ozawa) [inline WordPress/Classes/ViewRelated/Blog/Blog Dashboard/ViewModel/BlogDashboardViewModel.swift]:
```suggestion
            guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: cellType.defaultReuseID, for: indexPath) as? BlogDashboardCardConfigurable else {
                 return UICollectionViewCell()
            }
            cell.configure(blog: blog, viewController: viewController, dataModel: identifier.cellViewModel)
            return cell
```

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift
Comment (momo-ozawa) [inline WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift]:
```suggestion
    /// Posts are a special case: they might not be a 1-1 relation
```

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Posts/DashboardPostsCardCell.swift
Comment (momo-ozawa) [inline WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Posts/DashboardPostsCardCell.swift]: I wonder if there's solution that doesn't require "re-parsing" the data model again in the cell 🤔

**Review by momo-ozawa — COMMENTED** [review]
Works as described. This is a neat approach. 

A caveat is that the cellViewModel is a `NSDictionary?`, instead of a specific type, which means that we need to parse the view model when configuring the dashboard cells. I don't have any good solutions for this issue yet -- I'll let you know if I think of something.

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/ViewModel/BlogDashboardViewModel.swift
Comment (leandroalonso) [reply WordPress/Classes/ViewRelated/Blog/Blog Dashboard/ViewModel/BlogDashboardViewModel.swift]: One thing I'd like to avoid is returning a `UICollectionViewCell`. This shouldn't happen, for each registered card you need to provide a cell.

```suggestion
            let cell = collectionView.dequeueReusableCell(withReuseIdentifier: cellType.defaultReuseID, for: indexPath) as! BlogDashboardCardConfigurable
            cell.configure(blog: blog, viewController: viewController, dataModel: identifier.cellViewModel)
            return cell
```

What do you think of this approach?

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift
Comment (leandroalonso) [reply WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift:18]: Done in e2cd152ca4c23a1da7446a75e2db583c45b6a581

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/ViewModel/BlogDashboardViewModel.swift
Comment (leandroalonso) [reply WordPress/Classes/ViewRelated/Blog/Blog Dashboard/ViewModel/BlogDashboardViewModel.swift]: Done in 387089cc43fadd86250618b957cf518bed8c3c54

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Posts/DashboardPostsCardCell.swift
Comment (leandroalonso) [reply WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Posts/DashboardPostsCardCell.swift]: Hmm, this is a good idea. I'll improve that.

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift
Comment (momo-ozawa) [reply WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift:18]: Thanks! What do you think about the second part of my comment?  I was thinking we could iterate through remote cases / local cases individually, instead of looping through all cases then checking if it's remote or not

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Posts/DashboardPostsCardCell.swift
Comment (leandroalonso) [reply WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Posts/DashboardPostsCardCell.swift]: Changed in b81f7c7bc65c4e131dfc78e8b6112c579890b1c1

The final result is still not great tbh, but this connects directly with your comment related to the `NSDictionary`.

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift
Comment (leandroalonso) [reply WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift:18]: Oh, I missed that part, sorry.

The only I see is that this will create two groups: local cards will be displayed first, remote later. This is implicit while displaying with the `DashboardCard` order seems more natural to me. Wdyt?

**Comment by leandroalonso** [comment]
> A caveat is that the cellViewModel is a NSDictionary?, instead of a specific type, which means that we need to parse the view model when configuring the dashboard cells. I don't have any good solutions for this issue yet -- I'll let you know if I think of something.

@momo-ozawa 100% agree. I think this needs to be revisited because parsing the model in the cell configuration is not optimal and it's being done on the main thread.

What do you think about leaving that for another task? This PR is already big and we only have two models for now (posts and today stats). Then we can unblock all the other tasks but tackle that later.

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift
Comment (momo-ozawa) [reply WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift:18]: Right, I forgot about the ordering. Using the implicit ordering on DashboardCard sounds good.

**Comment by momo-ozawa** [comment]
> What do you think about leaving that for another task? This PR is already big and we only have two models for now (posts and today stats). Then we can unblock all the other tasks but tackle that later.

Sounds good!

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/ViewModel/BlogDashboardViewModel.swift
Comment (momo-ozawa) [reply WordPress/Classes/ViewRelated/Blog/Blog Dashboard/ViewModel/BlogDashboardViewModel.swift]: Sounds good -- I think we need cast to `DashboardCollectionViewCell` though. Otherwise we'll get a `Cannot convert return expression of type 'BlogDashboardCardConfigurable' to return type 'UICollectionViewCell?'` error

**Review by momo-ozawa — APPROVED** [review]
I left one [comment](https://github.com/wordpress-mobile/WordPress-iOS/pull/17909#discussion_r802534342) that requires a small change, but LGTM otherwise

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/ViewModel/BlogDashboardViewModel.swift
Comment (leandroalonso) [reply WordPress/Classes/ViewRelated/Blog/Blog Dashboard/ViewModel/BlogDashboardViewModel.swift]: Changed in 5c1f3cdee7b61528d87076c439b2d8aedb1d77c7 (ended up using `if let`), let me know wdyt.

Comment (momo-ozawa) [reply WordPress/Classes/ViewRelated/Blog/Blog Dashboard/ViewModel/BlogDashboardViewModel.swift]: 👍

**Pull request merged**
